### PR TITLE
fix(#1665): tabs add onChange event

### PR DESCRIPTION
**Pull request by syedszeeshan**
# Before (the change)

# After (the change)
-The _change event is available (onchange in React), returns the selected tab's index in an object `detail: { tab: _currentTab },`.
-Added a new test case

## Make sure that you've checked the boxes below before you submit the PR

- [x] I have read and followed the [setup steps](https://goa-dio.atlassian.net/wiki/spaces/DS/pages/3086385489/Contribution+-+web+content#%5BinlineExtension%5DCode-contributions)
- [x] I have created necessary unit tests
- [x] I have tested the functionality in both React and Angular.

## Steps needed to test

**Angular**
```
import { Component } from "@angular/core";

@Component({
  selector: "abgov-tabs",
  templateUrl: "./tabs.html",
})
export class TabsComponent {
  handleTabChange(event: Event) {
    const customEvent = event as CustomEvent;
    const tabIndex = customEvent.detail.tab;
    console.log(`Tab changed to ${tabIndex}`);
  }
}


<goa-tabs (_change)="handleTabChange($event)">
  <goa-tab>
    <div slot="heading">Profile</div>
    <p>
      <b>Profile:</b> Lorem ipsum dolor sit amet.
    </p>
  </goa-tab>
  <goa-tab>
    <div slot="heading">
      Review pending
      <goa-badge type="important" content="1"></goa-badge>
    </div>
    <p>
      <b>Review pending:</b> Lorem ipsum dolor sit amet, consectetur 
    </p>
  </goa-tab>
  <goa-tab>
    <div slot="heading">
      Completed
      <goa-badge type="midtone" content="1"></goa-badge>
    </div>
    <p>
      <b>Completed:</b> Lorem ipsum dolor sit amet, consectetur adipiscing elit
    </p>
    <p>New text</p>
  </goa-tab>
  <goa-tab heading="Test 1"> Testing </goa-tab>
</goa-tabs>
```

**React**
```
import { GoABadge, GoATab, GoATable, GoATabs } from "@abgov/react-components";

export default function Tabs() {
  function handleTabChange(tabIndex: number): void {
    console.log(`Tab changed to ${tabIndex}`);
  }

  return (
    <GoATabs initialTab={2} onChange={handleTabChange}>
      <GoATab heading="Profile">
        <p>
          <b>Profile:</b> Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </p>
      </GoATab>
      <GoATab
        heading={
          <>
            Review pending <GoABadge type="important" content="1" />
          </>
        }
      >
        <GoATable variant="relaxed">
          <thead>
            <tr>
              <th>First name</th>
              <th>Last name</th>
              <th>Comment</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>Person 1's first name</td>
              <td>Person 1's last name</td>
              <td>Waiting for a week</td>
            </tr>
            <tr>
              <td>Person 2's first name</td>
              <td>Person 2's last name</td>
              <td>Waiting for 2 weeks</td>
            </tr>
          </tbody>
        </GoATable>
      </GoATab>
      <GoATab
        heading={
          <>
            Completed <GoABadge type="midtone" content="1" />
          </>
        }
      >
        <p>
          <b>Completed:</b> Lorem ipsum dolor sit amet, consectetur  aliqua.
        </p>
        <p>New paragraph. Lorem ipsum dolor sit amet, consectetur adipiscing</p>
      </GoATab>
    </GoATabs>
  );
}

```

### CODE REVIEW

**Comment by syedszeeshan** [comment]
image

**Comment by lizhuomeng71** [comment]
In Angular, if I have the same example code, the onTabChange event will not fire. Should we also update the onTabChange method to align with Angular naming standards?

**Comment by syedszeeshan** [comment]
> In Angular, if I have the same example code, the onTabChange event will not fire. Should we also update the onTabChange method to align with Angular naming standards?

As discussed, change this selector: "abgov-paginate" as in the sample code.

I am looking into this "when initialTab="2" is set and I click on tab 2 for the first time, the onTabChange event fires"

File: libs/web-components/src/components/tabs/Tabs.svelte [hunk 1]
```diff
@@ -107,16 +110,27 @@
...
     if (!_tabsEl) return;
 
+    const previousTab = _currentTab;
+
     // prevent tab from exceeding limits
     _currentTab = clamp(tab, 1, _tabProps.length);
 
+    if (previousTab != _currentTab) {
+      _rootEl.dispatchEvent(
+        new CustomEvent("onTabChange", {
```
Comment (chrisolsen) [inline libs/web-components/src/components/tabs/Tabs.svelte:126]: Change `onTabChange` to `_change` to follow the event naming convention.

File: libs/web-components/src/components/tabs/Tabs.svelte [hunk 2]
```diff
@@ -107,16 +110,27 @@
...
+          composed: true,
+          bubbles: true,
+          detail: _currentTab,
```
Comment (chrisolsen) [inline libs/web-components/src/components/tabs/Tabs.svelte:129]: Pass an object containing the _currentTab value as the detail instead

```
...
detail: { tab: _currentTab },
```

File: libs/web-components/src/components/tabs/Tabs.svelte [hunk 3]
```diff
@@ -94,7 +94,10 @@
       link.setAttribute("data-testid", `tab-${index + 1}`);
       link.setAttribute("role", "tab");
       link.setAttribute("href", path + "#" + tabSlug);
-      link.addEventListener("click", () => setCurrentTab(index + 1));
+      link.addEventListener("click", (e) => {
+        e.preventDefault();
```
Comment (chrisolsen) [inline libs/web-components/src/components/tabs/Tabs.svelte:98]: Does `preventDefault` prevent the browsers url from getting the hash appended?

File: libs/react-components/src/lib/tabs/tabs.tsx [hunk 4]
```diff
@@ -14,14 +19,33 @@ export interface GoATabsProps {
...
+      };
+    }
+  }, [onTabChange]);
+
   return (
-    <goa-tabs
-      initialtab={initialTab}
-      data-testid={testId}
-    >
+    <goa-tabs ref={ref} initialtab={initialTab} data-testid={testId}>
```
Comment (chrisolsen) [inline libs/react-components/src/lib/tabs/tabs.tsx:48]: Is this formatting done intentionally or by your editor?

File: libs/react-components/src/lib/tabs/tabs.tsx [hunk 5]
```diff
@@ -14,14 +19,33 @@ export interface GoATabsProps {
...
+export function GoATabs({
+  initialTab,
+  children,
+  testId,
+  onTabChange,
+}: GoATabsProps): JSX.Element {
+  const ref = useRef<HTMLElement>(null);
+
+  useEffect(() => {
+    const element = ref.current;
```
Comment (chrisolsen) [inline libs/react-components/src/lib/tabs/tabs.tsx:34]: I would think that `ref.current` would also have to be in the useEffect's dependency array

**Comment by syedszeeshan** [comment]
image

Tests ran successfully after code changes

File: libs/web-components/src/components/tabs/Tabs.svelte (hunk 1 above)
Comment (syedszeeshan) [reply libs/web-components/src/components/tabs/Tabs.svelte]: changed to _change and onChange for React

File: libs/web-components/src/components/tabs/Tabs.svelte (hunk 2 above)
Comment (syedszeeshan) [reply libs/web-components/src/components/tabs/Tabs.svelte]: done

File: libs/web-components/src/components/tabs/Tabs.svelte (hunk 3 above)
Comment (syedszeeshan) [reply libs/web-components/src/components/tabs/Tabs.svelte]: I have removed this (this change was not needed)

File: libs/react-components/src/lib/tabs/tabs.tsx (hunk 4 above)
Comment (syedszeeshan) [reply libs/react-components/src/lib/tabs/tabs.tsx:48]: this is done by the editor

**Comment by lizhuomeng71** [comment]
this issue is verified fix on react and angular

File: libs/web-components/src/components/tabs/Tabs.svelte [hunk 6]
```diff
@@ -128,7 +139,6 @@
         el.focus();
       }
     });
-
```
Comment (chrisolsen) [inline libs/web-components/src/components/tabs/Tabs.svelte:131]: Keep the whitespace between blocks of logic.

File: libs/web-components/src/components/tabs/Tabs.svelte [hunk 7]
```diff
@@ -139,7 +149,6 @@
         }),
       );
     }
-
```
Comment (chrisolsen) [inline libs/web-components/src/components/tabs/Tabs.svelte:142]: Keep the whitespace between blocks of logic.

File: libs/web-components/src/components/tabs/Tabs.svelte [hunk 8]
```diff
@@ -107,16 +107,27 @@
   }
 
   function removeKeyboardEventListeners() {
-    // TODO: determine if this is ever being called
```
Comment (chrisolsen) [inline libs/web-components/src/components/tabs/Tabs.svelte:110]: did you determine if this is being called?

File: libs/web-components/src/components/tabs/Tabs.svelte (hunk 6 above)
Comment (syedszeeshan) [reply libs/web-components/src/components/tabs/Tabs.svelte]: done

File: libs/web-components/src/components/tabs/Tabs.svelte (hunk 7 above)
Comment (syedszeeshan) [reply libs/web-components/src/components/tabs/Tabs.svelte]: done

File: libs/web-components/src/components/tabs/Tabs.svelte (hunk 8 above)
Comment (syedszeeshan) [reply libs/web-components/src/components/tabs/Tabs.svelte:110]: Yes, I did a console log together with a breakpoint and it was hit.

**Review by chrisolsen — APPROVED** [review]

**Pull request merged**